core/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java

```diff
@@ -209,9 +222,24 @@ public Settings settings() {
      *
      * @param type   The mapping type
      * @param source The mapping source
+     * @deprecated use {@link #mapping(String, BytesReference, XContentType)}
      */
+    @Deprecated
     public PutIndexTemplateRequest mapping(String type, String source) {
-        mappings.put(type, source);
+        XContentType xContentType = XContentFactory.xContentType(source);
+        return mapping(type, new BytesArray(source.getBytes(StandardCharsets.UTF_8)), xContentType);
+    }
+
+    /**
+     * Adds mapping that will be added when the index gets created.
+     *
+     * @param type   The mapping type
+     * @param source The mapping source
+     * @param xContentType The type of content contained within the source
+     */
+    public PutIndexTemplateRequest mapping(String type, BytesReference source, XContentType xContentType) {
```

Comment:
why? the REST layer passes in the whole source, not just the mappings? Then the source is parsed and the mappings get converted to json in all cases, these mapping setters are just for the transport client I think.